### PR TITLE
[4.0][com_fields] Fix draggable sorting

### LINKED ISSUE #30073: [4.0] Custom field sorting - bottom only
### Steps to reproduce the issue

I have 5 Articles custom fields, 1 without a category, 1 and 3 in named categories. I need to change the order for front end purposes.

### Expected result

In the list of fields I should be able to drag a field up or down.

### Actual result

I can only drag a field to the bottom of the list.

### System information (as much as possible)

Firefox and Chrome on Mac OS X.

### Additional comments

Not too cumbersome with a list of 5 but would be unusable with a long list.

### PR DESCRIPTION
**Pull request by Quy**
Closes #30073.

### Summary of Changes
Fix draggable sorting

### Testing Instructions
Go to Content > Fields
Create fields
Click Ordering header
Drag rows to sort
Create field groups
Click Ordering header
Drag rows to sort
Assign fields to field groups
Drag rows to sort within field groups

### Actual result BEFORE applying this Pull Request
Sortable in last position only

### Expected result AFTER applying this Pull Request
Sortable in any positions

### CODE REVIEW

**Comment by ceford** [comment]
I have tested this item :red_circle: unsuccessfully on b44e874272572683b2d9d7a51b7c70bf235a32a8<br /><br />Can sort within field group but the sorting behaviour is still not as required. I have 5 fields, 1 not in a group, 1 and 3 in named groups. The individual items I can only drag to the bottom of the list. I believe I should be able to sort the individual items irrespective of their groups.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30094">issues.joomla.org/tracker/joomla-cms/30094</a>.</sub>

**Comment by Quy** [comment]
Is sorting an issue with Fields only or with other components too?

**Comment by Quy** [comment]
I can sort individually, but not an individual into a group. I assume that is the expected behavior.

**Comment by Quy** [comment]
Can you provide a screenshot of the fields so I can see what you are trying to do?

**Comment by ceford** [comment]
Here is a case where I have ordered fields crossing groups but I can only do it by dragging down.
image
I assumed that selecting order despite group is the expected behaviour.

**Comment by Quy** [comment]
For the first field, there are no other fields without a field group assigned to sort against.
Yes, it allows you to sort by dragging to the bottom, but essentially, it is still the same position.

For the `More Info` group, you can sort fields within this group without dragging to the bottom.

Click on the Filter Options.
Filter on `Flower data` group.
There is only one entry so no sorting possible.
Basically, this is the same scenario with `Latin name` field.

**Comment by ceford** [comment]
The sort order in the fields list affects the display order in the front end. Examples before and after changing order:
image

image

**Comment by Quy** [comment]
It could be related to this issue #23905. This PR fixes the dragging to the bottom only.

**Comment by Quy** [comment]
Is the latest screenshot a result of this PR?

**Comment by ceford** [comment]
No - the screenshots were without the patch. But with the patch it makes no difference - I can drag anything to the bottom of the list but only a group member upwards to within its group.

**Comment by brianteeman** [comment]
I would not expect to be able to sort into a group. Just as you can not sort into a category etc

**Comment by ceford** [comment]
> I would not expect to be able to sort into a group. Just as you can not sort into a category etc

But articles, users and contacts don't offer sorting by category!

**Comment by brianteeman** [comment]
oh yes they do

keyb

**Comment by ceford** [comment]
I am still baffled! With the patch applied, referring to my illustration above, I can drag any member of the More Info group up or down within its group. I can drag any field to the bottom of the list as illustrated. If I drag the ungrouped item or the Flower data group item to the bottom of the list I can't drag it up again. Is this the desired behaviour? Or even desirable behaviour? 

I am revising the Help screens for Fields and Field Groups so I really need to understand it all myself. As I see it, the Field Group order just controls the Tab order in the data entry form and has no effect in the front end. The Field order affects the order within the data entry tab and the display order in the front end. Is that right?

**Comment by ceford** [comment]
By way of illustration, I just removed the patch and decided to put the Fields in the correct order. I had to move each one individually to the bottom of the list. Not too tedious with 5, but imagine doing it with 10 or 20 and making a mistake.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30094">issues.joomla.org/tracker/joomla-cms/30094</a>.</sub>

**Comment by brianteeman** [comment]
> As I see it, the Field Group order just controls the Tab order in the data entry form and has no effect in the front end.

No. It also controls the order in the front end

>  The Field order affects the order within the data entry tab and the display order in the front end. 

Yes

**Comment by ceford** [comment]
> > As I see it, the Field Group order just controls the Tab order in the data entry form and has no effect in the front end.
> 
> No. It also controls the order in the front end

That seems to be not the case. In a fresh test these are my field groups:

image

And the Latin name field is not in a group so appears in the Fields tab. This is my Fields order:

image

And this is the front end rendering:

image

So the front-end order seems to be controlled only by the Field order.

> 
> > The Field order affects the order within the data entry tab and the display order in the front end.
> 
> Yes

Agree!

**Comment by Quy** [comment]
> I had to move each one individually to the bottom of the list. Not too tedious with 5, but imagine doing it with 10 or 20 and making a mistake.

Use Filter Options to select the desired field group to display applicable fields. With this PR, you can drag and drop within the group and not to the bottom only. 

In other words, filter the list to exclude fields not relevant to the field group so you don’t have to interact with them.

**Comment by ceford** [comment]
> Use Filter Options to select the desired field group to display applicable fields. With this PR, you can drag and drop within the group and not to the bottom only.
> 
> In other words, filter the list to exclude fields not relevant to the field group so you don’t have to interact with them.

That does not make sense. The Select Field Group drop-down list does not contain No Field Group and if I select Flower Data I cannot move it with respect to More Data. So I repeat: the order shown in the Fields list is the order displayed in the front end. We really need to be able to move any field up or down irrespective of field group.

**Comment by brianteeman** [comment]
> We really need to be able to move any field up or down irrespective of field group.

No we dont - that makes no sense and defeats the objective of grouping the fields.

> the order shown in the Fields list is the order displayed in the front end.

And that is correct.

With this PR the code is working correctly as designed and exactly how it is working in joomla 3

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 7bc66cc6714ecdc5b66524876c7982aeb6d8b8bf<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30094">issues.joomla.org/tracker/joomla-cms/30094</a>.</sub>

**Comment by bonzani** [comment]
I have tested this item :white_check_mark: successfully on 7bc66cc6714ecdc5b66524876c7982aeb6d8b8bf<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30094">issues.joomla.org/tracker/joomla-cms/30094</a>.</sub>

**Comment by ceford** [comment]
OK - I rethink my use of Fields and Field Groups to use them as designed, and document accordingly. Sorry for being such a pain!

**Comment by jan-klima** [comment]
Hey guys, I know its old but it still doesnt work properly for me. I am on J4.1.5 and I can only drag all the way down + it doesnt save it. Resets position on reload...
